File: Reqnroll/Chapter15/15.7-TheStubDependencyPattern/Before/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(OrderingApiDriver orderingApiDriver, NotificationsApiDriver notificationsApiDriver)
{
    public record OrderRequestData(TimeSpan ExpectedDeliveryTime);

    [Given("they have placed an order")]
    public async Task GivenTheyHavePlacedAnOrder(DataTable dataTable)
    {
        var orderData = dataTable.CreateInstance<OrderRequestData>();
        var expectedDeliveryTime = TimeOnly.FromTimeSpan(orderData.ExpectedDeliveryTime);

        // With the real time service we cannot fast-forward time, so cannot use the specified
        // expectedDeliveryTime. Instead, we force the expected delivery time being in 0.5 seconds,
        // and we wait in the WhenTheDeliveryHasNotBeenMadeBy method for the background timer loop
        // to process the subscription.
        var placeOrderRequest = new PlaceOrderRequestObjectMother()
            .WithExpectedDeliveryTime(DateTimeOffset.Now.AddSeconds(0.5))
            .Build();

        await orderingApiDriver.PlaceOrder(placeOrderRequest).Execute();
    }

    [When("the delivery has not been made by {TimeOnly}")]
    public void WhenTheDeliveryHasNotBeenMadeBy(TimeOnly time)
    {
        //Workaround: see notes above!
        Thread.Sleep(TimeSpan.FromMilliseconds(1500));
    }

    [Then("the customer should receive a notification about the delay")]
    public async Task ThenTheCustomerShouldReceiveANotificationAboutTheDelay()
    {
        var notifications = await notificationsApiDriver.GetNotifications(DomainDefaults.CustomerName).Execute();

        Assert.IsNotNull(notifications);
        Assert.IsTrue(notifications.Any(n =>
                n.Message.Contains("delayed", StringComparison.OrdinalIgnoreCase)),
            "Expected a delay notification but none was found.");
    }
}

File: CucumberJVM/Chapter16/16.3-EnsurePattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.specs.drivers.OrderingApiDriver;
import com.wimp.app.specs.support.OrderingContext;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

public class OrderingStepDefinitions {
    private final OrderingContext orderingContext;
    private final OrderingApiDriver orderingApiDriver;

    public OrderingStepDefinitions(OrderingContext orderingContext, OrderingApiDriver orderingApiDriver) {
        this.orderingContext = orderingContext;
        this.orderingApiDriver = orderingApiDriver;
    }

    @Given("they have placed an order for {int} pizzas")
    public void theyHavePlacedAnOrderForPizzas(int count) throws Exception {
        var placeOrderRequest = new PlaceOrderRequestObjectMother().withItems(count).build();
        var placedOrder = orderingApiDriver.placeOrder(placeOrderRequest).execute();
        orderingContext.setCurrentOrderNo(placedOrder.getOrderNo());
    }
}
